zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,195 @@
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     * @param converter a function to convert zNode value to a List of {@link Endpoint}
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout,
+                                  ZkNodeValueConverter converter) {
+
+        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        this.zNode = requireNonNull(zNode, "zNode");
+        this.converter = requireNonNull(converter, "converter");
+        try {
+            zooKeeper = new ZooKeeper(zkConnectionStr, sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    connectionLatch.countDown();
+                }
+            });
+            connectionLatch.await();
+            Stat stat = zooKeeper.exists(zNode, new ZkWatcher());
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = this.converter.convert(nodeData);
+            }
+        } catch (Exception e) {
+            throw new EndpointGroupException(
+                    "failed to connect to ZooKeeper:  " + zkConnectionStr + " with error:  " + e.getMessage());
+        }
+    }
+
+    /**
+     * Create a Zookeeper endpoint group with a {@link DefaultZkNodeValueConverter}.
+     * @param zkConnectionStr A connection string containing a comma
+     *                          separated list of host:port pairs , each corresponding to a ZooKeeper server
+     * @param zNode a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout   Zookeeper session timeout in milliseconds
+     */
+    public ZookeeperEndpointGroup(String zkConnectionStr, String zNode, int sessionTimeout) {
+        this(zkConnectionStr, zNode, sessionTimeout, new DefaultZkNodeValueConverter());
+    }
+
+    @Override
+    public List<Endpoint> endpoints() {
+        return endpointList;
+    }
+
+    final class ZkWatcher implements Watcher, StatCallback {
+        private int retryDelay = 1000; //start with one second
+
+        @Override
+        public void process(WatchedEvent event) {
+            String path = event.getPath();
+            if (event.getType() != Event.EventType.None) {
+                if (path != null && path.equals(zNode)) {
+                    // Something has changed on the node, let's find out
+                    zooKeeper.exists(zNode, true, this, null);
+                }
+            }
+        }
+
+        @Override
+        public void processResult(int responseCodeInt, String path, Object ctx, Stat stat) {
+            boolean exists;
+            Code responseCode = get(responseCodeInt);
+            switch (responseCode) {
+                case OK:
+                    exists = true;
```

Comment:
it is a local variable ,every time this method is invoked(It is callback) ,it will be reset to the default value    `false`